doc/developer/design/20230607_shift_status_updates.md

```diff
@@ -0,0 +1,151 @@
+working on things that are crossing team boundaries it will be helpful to spell
+out any new interfaces/traits/interactions.
+
+For most new features, you should think about testing, rollout/lifecycle, and
+observability. These things can warrant their own sections.
+-->
+
+We will introduce a couple new pieces of state to `mz_storage::storage_state::StorageState` in order to keep track of the statuses of sink and source ingestion dataflows:
+
+```rust
+pub source_statuses: Rc<RefCell<BTreeMap<GlobalId, Option<HealthStatus>>>>,
+pub sink_statuses: Rc<RefCell<BTreeMap<GlobalId, Option<HealthStatus>>>>,
+```
+
+For source ingestion dataflows, the health operator will evolve to update the status corresponding to the source's `GlobalId` in `source_statuses`. For sinks, `Healthchecker::update_status` will evolve to do the same for `sink_statuses`.
+
+The storage worker will be updated to periodically send `StorageResponse::StatusUpdates` to the storage controller, which will be defined as:
+
+```rust
+pub enum StorageResponse {
+    ...,
+    StatusUpdates(Vec<ObjectStatusUpdate>),
+}
+
+pub enum ObjectStatusUpdate {
+    Sink(SinkStatusUpdate),
+    Source(SourceStatusUpdate),
+}
+
+// Same for `SinkStatusUpdate`
+pub struct SourceStatusUpdate {
+    id: GlobalId,
+    status: String,
+    error: String,
+    hint: Option<String>,
+}
+```
+
+To avoid sending this information on every tick of the storage worker's main processing loop, we will only send updates per some defined frequency (e.g. every 5 seconds). This is similar to how statistics reporting is currently implemented. Upon receiving a `StorageResponse::StatusUpdates` message, the storage controller will append the status updates to the appropriate managed collection. Writes to the backing persist shard will take place in a dedicated task, namely the one spawned by `mz_storage_client::controller::collection_mgmt::CollectionManager`.
+
+When dropping a replica, the storage controller will need to retrieve all source/sink dataflows that have been observed to be associated with the given storage instance. While this state is currently maintained in `mz_storage_client::controller::rehydration::RehydrationTask`, we can lift it up into `RehydrationStorageClient` in order for the storage controller to access it without duplicating the data. Upon dropping an instance, the storage controller can be sure that it won't receive any more `StorageResponse`'s from that instance. As a result, late status updates after an instance is dropped are guaranteed not to occur.
+
+Immediately after the coordinator announces initialization is complete, the storage controller will be updated to look for any ingestion/export dataflows that belong to an instance with 0 replicas and set their status to `paused`. This should allow `environmentd` to correctly set the status for sources/sinks in scenarios where it crashes before it was able to write the `paused` status to persist.
```

Comment:
One related point that came up in the first implementation: when a `clusterd` is unhealthy enough that a source or sink is unable to write its status, the stored status will be similarly misleading. (After this change, problems with network connectivity between `environmentd` and `clusterd` will also trigger this issue.)

Thoughts on adding either a new status, or overloading an existing status, to report that a source or sink is in an unknown state?

Reply:
Oh, that's a great idea! If the storage controller loses contact with a replica I'd love if the controller wrote a status of `unknown` to the table.

Reply:
Good point! I do think that we can get away without the controller having to write a new status upon losing connection with a replica. An `unknown` status can be deduced by looking at the time of the last status update for a source/sink. We could probably handle that in the view sql for `mz_{source|sink}_statuses`. For example if the last update was over X minutes ago, display a status of `unknown` with reason `"last reported status over X minutes ago"`. Does that sound reasonable?

Reply:
Today's status update table normally only has an entry when the status changes. If I understand your suggestion, it implies adding a new entry every X seconds, even if the status has not changed. Under reasonable assumptions, that probably increases the size of the table by something like 1000x. (Which means that any queries against that table will have 1000x as much work to do, etc.)

I'd have to run some tests to know whether that's a total dealbreaker or not, but my gut is that the benefits of not having to write the new status are outweighed by the costs in any case. Do let me know if I'm missing something however!

Reply:
Ahh, right! I had totally forgot that when writing my comment 😆.

With that in mind I agree that we should write an `unknown` status if the response channel from the replica is broken